### PR TITLE
logging: log_backend_swo: Add support for logging v2

### PR DESCRIPTION
**Pull request by nordic-krch**
Added support for logging v2 backend API for SWO backend.

### CODE REVIEW

**Review by dcpleung — APPROVED** [review]

File: subsys/logging/log_backend_swo.c [hunk 1]
```diff
@@ -144,10 +152,11 @@ static void log_backend_swo_sync_hexdump(
...
 
 const struct log_backend_api log_backend_swo_api = {
-	.put = IS_ENABLED(CONFIG_LOG_IMMEDIATE) ? NULL : log_backend_swo_put,
-	.put_sync_string = IS_ENABLED(CONFIG_LOG_IMMEDIATE) ?
+	.process = IS_ENABLED(CONFIG_LOG2) ? log_backend_swo_process : NULL,
+	.put = IS_ENABLED(CONFIG_LOG_MODE_DEFERRED) ? log_backend_swo_put : NULL,
+	.put_sync_string = IS_ENABLED(CONFIG_LOG_MODE_IMMEDIATE) ?
 			log_backend_swo_sync_string : NULL,
-	.put_sync_hexdump = IS_ENABLED(CONFIG_LOG_IMMEDIATE) ?
+	.put_sync_hexdump = IS_ENABLED(CONFIG_LOG_MODE_IMMEDIATE) ?
```
Comment (mnkp) [inline subsys/logging/log_backend_swo.c:159]: This and updates to `.put`, `.put_sync_string` look like a bugfix, an unrelated change. This should likely go into a separate commit.

**Review by mnkp — COMMENTED** [review]
This PR contains some (minor) changes unrelated to adding support for v2 logging in SWO backend. If we want to keep them in this PR the PR description should be updated.

File: subsys/logging/log_backend_swo.c (hunk 1 above)
Comment (nordic-krch) [reply subsys/logging/log_backend_swo.c:159]: it's not a bugfix, without v2 support using `IS_ENABLED(CONFIG_LOG_IMMEDIATE)` was ok but `CONFIG_LOG_IMMEDIATE` is set for both immediate versions so now we need better distinction and thus `CONFIG_LOG_MODE_IMMEDIATE` is used which is v1 immediate mode (with `CONFIG_LOG2_MODE_IMMEDIATE` for v2)

**Review by mnkp — APPROVED** [review]
